index_alias_impl.go

```diff
@@ -866,3 +856,45 @@ func (f *indexAliasImplFieldDict) Close() error {
 	defer f.index.mutex.RUnlock()
 	return f.fieldDict.Close()
 }
+
+func finalizePreSearchResult(req *SearchRequest, preSearchResult *SearchResult) {
+	if requestHasKNN(req) {
+		preSearchResult.Hits = finalizeKNNResults(req, preSearchResult.Hits)
+	}
+}
+
+// -----------------------------------------------------------------------------
+// KNNPreSearchResultProcessor is a preSearchResultProcessor that
+// processes the preSearchResult for KNN queries. It adds the KNN hits
+// to the KNNCollectorStore, and then calls the Final method to get the
+// final KNN hits.
+type KnnPreSearchResultProcessor struct {
+	Add      func(*SearchResult, string)
+	Finalize func(*SearchResult)
+}
+
+type PreSearchResultProcessor struct {
+	knnProcessor *KnnPreSearchResultProcessor
+}
+
+func CreatePreSearchResultProcessor(req *SearchRequest) *PreSearchResultProcessor {
```

Comment:
Why can't return type here simply be KnnPreSearchResultProcessor? I don't really see the need for 2 structs above. We can drop `PreSearchResultProcessor`.

Reply:
the aim is to allow `Presearch` to be generic construct that can be reused for other purposes. Lets say i have another logic for merging synonym results from multiple partitions, i would add the `SynonymPreSearchResultProcessor` as another struct in the PreSearchResult Processor, so any presearch result containing either `knn` or `synonym` data can be processed independently. By just keeping it as `KnnPreSearchResultProcessor` in `PreSearchDataSearch`, i would have to refactor it later on for synonym search.

Reply:
I understand your rationale but we can adapt this or add a separate processor for synonyms later, we'll also want to refactor the name - `preSearchDataSearch` at that point as well, which is quite distasteful.

For now let's reduce the number of redirects, keep it simple.

Reply:
have made it an interface and one struct implementing it now ... cannot use it directly as the index_alias_impl.go file is not in the vectors tag so it cannot read the knn specific code like the knn store ...